src/Compilers/CSharp/Portable/Syntax/SyntaxExtensions.cs

```diff
@@ -387,5 +387,43 @@ internal static bool IsOutVarDeclaration(this DeclarationExpressionSyntax p)
         {
             return p.Designation.Kind() == SyntaxKind.SingleVariableDesignation && p.IsOutDeclaration();
         }
+
+        /// <summary>
+        /// Visits all the subTypeSyntaxes of a typeSyntax in depth first order, invoking an action on each one in turn, before finally invoking the action on the passed in TypeSyntax
+        /// </summary>
+        /// <param name="type"></param>
+        /// <param name="action"></param>
+        internal static void VisitTypeSyntaxes(this TypeSyntax type, Action<TypeSyntax> action)
+        {
+            switch (type)
+            {
+                case ArrayTypeSyntax arrayTypeSyntax:
+                    arrayTypeSyntax.ElementType.VisitTypeSyntaxes(action);
+                    break;
+                case NullableTypeSyntax nullableTypeSyntax:
+                    nullableTypeSyntax.ElementType.VisitTypeSyntaxes(action);
+                    break;
+                case PointerTypeSyntax pointerTypeSyntax:
+                    pointerTypeSyntax.ElementType.VisitTypeSyntaxes(action);
+                    break;
+                case TupleTypeSyntax tupleTypeSyntax:
+                    foreach (var element in tupleTypeSyntax.Elements)
+                    {
+                        element.Type.VisitTypeSyntaxes(action);
+                    }
+                    break;
+                case RefTypeSyntax refTypeSyntax:
+                    refTypeSyntax.Type.VisitTypeSyntaxes(action);
+                    break;
+                case GenericNameSyntax genericNameSyntax:
+                    foreach (var typeArgument in genericNameSyntax.TypeArgumentList.Arguments)
+                    {
+                        typeArgument.VisitTypeSyntaxes(action);
+                    }
+                    break;
+            }
```

Comment:
>} [](start = 12, length = 1)

Should we handle ```QualifiedNameSyntax``` as well? I think we would want to add a ```default:``` case that throws ``ExceptionUtilities.Unreachable```. 

Also, I think would be better to make this method more specialized and perform call back only for ```ArrayTypeSyntax.RankSpecifiers```. None of the consumers are interested to be called for anything else.